compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,148 @@
+// are cached in hoistableOpMap.
+//
+// It expects the op's producers have been checked and the results exist in
+// `hoistableOpMap`.
+static bool
+isHoistableOp(LoopLikeOpInterface loopOp, Operation *op,
+              llvm::SmallDenseMap<Operation *, bool> &hoistableOpMap) {
+  // Currently we don't handle implicit captures, so don't hoist ops with
+  // regions.
+  if (op->getNumRegions() > 0) {
+    hoistableOpMap[op] = false;
+    return false;
+  }
+
+  // Check if the op type is hoistable.
+  if (!isa<tensor::EmptyOp, tensor::PackOp>(op)) {
+    hoistableOpMap[op] = false;
+    return false;
+  }
+
+  // Check if all producers are hoistable.
+  bool hoistable = true;
+  for (OpOperand &operand : op->getOpOperands()) {
+    Value value = operand.get();
+    // Ignore values defined outside the loop.
+    if (loopOp.isDefinedOutsideOfLoop(value))
+      continue;
+
+    Operation *producer = value.getDefiningOp();
+    // If the producer is not an operation, can't hoist it.
+    if (!producer) {
+      hoistable = false;
+      break;
+    }
+
+    auto producerResult = hoistableOpMap.find(producer);
+    assert(producerResult != hoistableOpMap.end() &&
+           "producer should have been checked");
+    if (!producerResult->second) {
+      hoistable = false;
+      break;
+    }
+  }
+
+  hoistableOpMap[op] = hoistable;
+  return hoistable;
+}
+
+static LogicalResult hoistLoopInvariants(LoopLikeOpInterface loopOp,
+                                         RewriterBase &rewriter) {
+  // First find the root ops can be hoisted. The root op needs to satisfy:
+  // 1. It is a root op having benefits to be hoisted (e.g. tensor.pack)
+  // 2. Its backward slice can be hoisted (e.g. they are loop invariant)
+  SmallVector<Operation *> opsToHoist;
+  llvm::SmallDenseMap<Operation *, bool> hoistableOpMap;
+  for (Region *region : loopOp.getLoopRegions()) {
+    // Consider only the top-level ops in the region.
+    // The forward visiting also ensures we are check and add ops in topological
+    // order.
+    for (Operation &op : region->getOps()) {
+      if (isHoistableOp(loopOp, &op, hoistableOpMap)) {
+        LLVM_DEBUG(LICM_DBGS() << "Found hoistable op: " << op << "\n");
+        opsToHoist.push_back(&op);
+      }
+    }
+  }
+  if (opsToHoist.empty())
+    return success();
+
+  // Wrap the loop in zero-trip-check so the hoisted ops will only run when the
+  // loop condition is ever satisfied.
+  FailureOr<LoopLikeOpInterface> wrappedLoop =
+      TypeSwitch<Operation *, FailureOr<LoopLikeOpInterface>>(
+          loopOp.getOperation())
+          .Case<scf::WhileOp>([&](scf::WhileOp op) {
+            return scf::wrapWhileLoopInZeroTripCheck(op, rewriter);
+          })
+          .Default([&](Operation *op) { return failure(); });
+  if (failed(wrappedLoop))
+    return failure();
+
+  // Hoist ops in their original insertion order, which is the topological
+  // order.
+  for (Operation *op : opsToHoist) {
```

Comment:
I am surprised this works. You need to use a `MapVector` to ensure that things are in insertion order. `SmallDenseMap` does not give you any gaurantees (and  can cause intermittent failures).

Reply:
`opsToHoist` is a `SmallVector` but, yes, we can turn it into a `SetVector` and remove the other map we use.

Reply:
> `opsToHoist` is a `SmallVector` but, yes, we can turn it into a `SetVector` and remove the other map we use.

Ah ok, I missed that part, but SetVector seems simpler.

Reply:
Changed to SetVector